src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleObject.cs

```diff
@@ -30,18 +38,6 @@ private static void HandleStartObject(JsonSerializerOptions options, ref ReadSta
 
             JsonClassInfo classInfo = state.Current.JsonClassInfo;
 
-            if (classInfo.CreateObject is null && classInfo.ClassType == ClassType.Object)
```

Comment:
No, there's no functional change for `ClassType.Object` types here.

The problem with the way this check was written is that it assumes `classInfo.CreateObject` being `null` is only a problem when we are processing `ClassType.Object`. This led to null refs otherwise.

The checks here have been refactored to express that, after pushing the property being examined to the read stack (if necessary), only `ClassType.Dictionary` and `ClassType.Object` types are valid when we see a `StartObject` token. 

`ClassType.Value` and `ClassType.Unknown` are handled (with converters) in https://github.com/dotnet/corefx/blob/4c3bd8a0533a2e35df371b91045a1a5f0ada7d54/src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs#L63-L70.

If we have `ClassType.Enumerable`, this means we have invalid Json, and we throw `JsonException`.